examples/run_language_modeling.py

```diff
@@ -144,8 +146,86 @@ def __getitem__(self, i):
         return torch.tensor(self.examples[i], dtype=torch.long)
 
 
+class LazyLineByLineTextDataset(Dataset):
+    """
+    Credit: @bramvanroy for this linecache implementation.
+    """
+
+    def __init__(self, file_path):
+        self.file_path = file_path
+        self.num_entries = self._get_n_lines(self.file_path)
+
+    @staticmethod
+    def _get_n_lines(file_path):
+        with Path(file_path).resolve().open(encoding="utf-8") as fhin:
+            for line_idx, _ in enumerate(fhin, 1):
+                pass
+
+        return line_idx
+
+    def __getitem__(self, idx):
+        """
+
+        :param idx (int): the index of the line to get
+        :return (str or None): The line as a string (newline removed) or None if there is an exception.
+        """
+        # linecache starts counting from one, not zero, +1 the given index
+        line = linecache.getline(self.file_path, idx + 1).strip()
+        return line
+
+    def __len__(self):
+        return self.num_entries
+
+
+def make_collate(tokenizer, block_size, lazy=False):
+    """
+    Constructs collate function for Dataloader.
+
+    :param tokenizer: tokenizer object (instantiated) used to tokenize lines of text.
+    :param block_size: (int) tokenized sequences truncated so as to not exceed this length.
+    :param lazy: (bool) whether or not to use return a collate function for lazy text loading.
+    :return: (fn: list[str or torch.tensor[int] (1d)] -> torch.tensor[int] (2d)
+    """
+
+    def lazy_collate(examples):
+        """
+        Dataloader collate_fn for Lazy data loading.
+
+        :param examples: (list[str]) the text lines to be collated in this batch.
+        :return: (torch.tensor) of the tokenized examples padded/truncated to form a 2d tensor of ints.
+        """
+        # Filter empty strings. LazyLineByLineTextDataset will return empty string if there is an error on a line.
+        examples = [ex for ex in examples if ex]
+        examples = tokenizer.batch_encode_plus(examples, max_len=block_size)
+        # Seems that the above tokenisation statement doesn't correctly truncate leading to sequence-too-long errors in
+        # training.
+        examples = [torch.tensor(ex[:block_size]) for ex in examples["input_ids"]]
```

Comment:
I guess you could just use `tokenizer.batch_encode_plus(examples, max_len=block_size, return_tensors="pt")`

Reply:
Just realised: I don't think this will work for GPT architectures. Just tried it on openai-gpt and get.

```
ValueError: Unable to set proper padding strategy as the tokenizer does not have a padding token. In this case please set the `pad_token` `(tokenizer.pad_token = tokenizer.eos_token e.g.)` or add a new pad token via the function add_special_tokens if you want to use a padding strategy
```

There was this kind of logic before to deal with this 
```python
if tokenizer._pad_token is None:
    return pad_sequence(examples, batch_first=True)
return pad_sequence(examples, batch_first=True, padding_value=tokenizer.pad_token_id)
```

Is there a way of nicely coping with this `batch_encode_plus`? Or do we have to fall back to `torch.nn.utils.rnn.pad_sequence`?

Reply:
@patrickvonplaten @thomwolf We discussed this on Slack but I'm not sure what the best approach is here: what should we do padding-wise with the models that do not have a padding token? Just pad with zeroes (even though that is not the padding id) which should then be resolved by the causal masking anyway?

Reply:
@julien-c Everything LGTM but the comment above must be addressed first: how to deal with tokenizers that do not have apdding token.

Reply:
What i would do here is to simply not allow padding when there is no `pad_token_id`. As far as I know with the fast tokenizers, it will not be possible anymore to pad if there is no `pad_token_id` anymore (@mfuntowicz). So as the error message states above the user should add the `pad_token_id` himself (He could for example just set it to 0 and be aware that he has to set `lm_labels` correctly. to not calculate a loss on the padded tokens. Is there a big disadvantage to force the user to set the pad_token_id himself besides the "less-friendly" user-experince? 
Doing something hacky here in the code is not a very safe in my opinion and could easily lead to silent bugs in the future.

Reply:
@patrickvonplaten Perhaps a separate PR is needed before this one gets merged, I guess? The problem is that you say that the "user must set the pad token ID manually", does that mean that we should add an optional argument to the script that is only required when the chosen model does not have a padding token?

Reply:
I would actually say that this can be merged without any special code that sets a `pad_token_id`. As far as I can see the tokenizer will be an argument to the function no? So if the user calls the function with a tokenizer and GPT2, an error pops up and the user knows he has to set a `pad_token_id` for the tokenizer and then insert this tokenizer to the function again. We could still make it easier for the user afterward without breaking backward compatibility

Reply:
In the script, yes, but since this is part of the `run_language_modeling.py` script, wouldn't we want people to be able to run the script without changing the code but only changing the command line arguments?

Reply:
I see! In that case, we could add a `force_padding` argument or something. When doing this we have to be sure that the `lm_labels` are correctly set to `-100` for all padded tokens.

 I think the best approach here is to
1) Force some token to be the padding token, e.g. 0 
2) set `tokenizer.pad_token_id = 0` 
3) pad the input batch with `batch_encode_plus()`. 
4) Retrieve the `attention_mask` 
5) Use the `attention_mask` to set `lm_labels = -100`. 

**Note**: In the case when the tokenizer does not have a `pad_token_id`, we cannot simply set `pad_token_id = 0` and set all `lm_labels = -100` where `input_ids = 0` because the `token_id 0` can actually exist and its `lm_labels` would then incorrectly be set to `-100`. So we have to use the `attention_mask` created by `batch_encode_plus.

Reply:
@patrickvonplaten Is this something you can implement for the run_language_modeling.py script? @julien-c I am not sure whether we should first merge this commit (even though using a model without padding token will cause issues) and then make the other changes that Patrick suggests, or the other way around.

Reply:
I tried to implement @patrickvonplaten 's idea in the latest commit. It runs fine for GPT-2 and OpenAI-GPT. I confess to checking only that it trains, not that it trains well.